File: java-playground/src/test/java/org/gonza/javaplayground/car/RacingTest.java
package org.gonza.javaplayground.car;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class RacingTest {
    @Test
    void Racing은_Car를_생성할_수_있다() {
        List<String> names = List.of("test1", "test2", "test3");

        Racing racing = new Racing(names);

        assertEquals(names.size(), racing.countPlayers());
    }

    @Test
    void Car의_이름이_6자를_초과하면_예외가_발생한다() {
        List<String> names = List.of("123456");

        assertThrows(IllegalArgumentException.class, () -> new Racing(names));
    }

    @Test
    void 숫자가_4_이상인_Car만_이동한다() {
        Racing racing = new Racing(List.of("test1", "test2"));

        RaceResult result = racing.race(List.of(3, 4));

        assertEquals(1, result.findWinner().getFirst().getPosition());
        assertEquals("test2", result.findWinner().getFirst().getName());
    }

    @Test
    void Racing의_Car는_숫자_4_이상이면_이동한다() {
        Racing racing = new Racing(List.of("test1"));

        RaceResult result = racing.race(List.of(4));

        assertEquals(1, result.findWinner().getFirst().getPosition());
    }

    @Test
    void Racing의_Car는_숫자_4_미만이면_이동하지_않는다() {
        Racing racing = new Racing(List.of("test1"));

        RaceResult result = racing.race(List.of(3));

        assertEquals(0, result.findWinner().getFirst().getPosition());
    }

    @Test
    void 결과를_출력할_수_있다() {
        Racing racing = new Racing(List.of("test1", "test2"));

        RaceResult firstResult = racing.race(List.of(4, 4));
        assertEquals("test1 : -\ntest2 : -", firstResult.report());

        RaceResult secondResult = racing.race(List.of(4, 4));
        assertEquals("test1 : --\ntest2 : --", secondResult.report());

        RaceResult thirdResult = racing.race(List.of(3, 4));
        assertEquals("test1 : --\ntest2 : ---", thirdResult.report());
    }
}

File: kotlin-playground/src/test/kotlin/org/gonza/kotlinplayground/domain/racing/RacingTest.kt
package org.gonza.kotlinplayground.domain.racing

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows

class RacingTest {
    val carsName = "test1,test2,test3"
    val round = 10
    @Test
    fun `경주는 텍스트값을 받아 생성된다`() {
        assertDoesNotThrow { Racing(carsName = carsName, round = round) }
    }

    @Test
    fun `경주차는 carsName의 쉼표를 기준으로 구분된다`() {
        val result = Racing(carsName = carsName, round = round)

        assertThat(result.carList.size).isEqualTo(3)
        assertThat(result.carList[0].name).isEqualTo("test1")
        assertThat(result.carList[1].name).isEqualTo("test2")
        assertThat(result.carList[2].name).isEqualTo("test3")
    }

    @Test
    fun `경주차 중 하나의 이름이 5자를 넘길 경우 IllegalArgumentException이 발생한다`() {
        val carsName="test1,over5test,test3"
        assertThrows<IllegalArgumentException> { Racing(carsName = carsName, round = round) }
    }

    @Test
    fun `경주는 입력받은 만큼의 라운드를 가진다`() {
        val result = Racing(carsName = carsName, round = round)
        assertThat(result.round.value).isEqualTo(round)
    }

    @Test
    fun `라운드는 0보다 작을 수 없다`() {
        val round = -1
        assertThrows<IllegalArgumentException> { Racing(carsName = carsName, round = round) }
    }
}